src/librustc_const_eval/check_match.rs

```diff
@@ -486,7 +486,7 @@ fn check_legality_of_move_bindings(cx: &MatchVisitor,
                              "cannot bind by-move with sub-bindings")
                 .span_label(p.span, "binds an already bound by-move value by moving it")
                 .emit();
-        } else if has_guard {
+        } else if has_guard && !cx.tcx.sess.features.borrow().bind_by_move_pattern_guards {
```

Comment:
Hmm, I think that this doesn't suffice. Specifically, there are cases we want to disallow, such as this:

```rust
let x = Some(vec![1, 2, 3]);
match x {
    Some(v) if { mem::drop(v); true } => { println!("empty"); }
    Some(v) => { println!("not-empty: {:?}", v); }
    None => { println!("None!"); }
}
```

The danger is that `v` the vector will get freed after the first guard is executed.

I think that the RFC specifically stated that we should allow this but only if the guard only uses the value "by reference" (or if it only copies). 

One way to do this analysis would be to deploy the `ExprUseVisitor`. I think the option I prefer at this point is to say that, when the guard executes, we treat the bound values as if there was an implicit shared reference (much like how a closure works). This probably requires an RFC, though I think it's backwards compatible.

In the example above, that would mean that references to `v`  in the pattern guard would be treated as equivalent to `*v0` where `v0` is a synthetic variable with type `&Vec<i32>`; hence the `mem::drop(v)` guard would be equivalent to `mem::drop(*v0)`, which would mean that it'd be an error (since you are moving through a borrowed reference).